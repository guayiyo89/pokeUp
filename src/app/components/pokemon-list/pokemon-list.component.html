<p>pokemon-list works!</p>

<input
  type="text"
  id="searchPokemon"
  [formControl]="searchPokemon"
  (keyup.enter)="showPokemon()"
/>
<button (click)="showPokemon()">Search</button>

<div *ngIf="load">
  <div *ngIf="pageActual.length > 0">
    <p *ngFor="let pokemon of pageActual">
      {{ pokemon.name }}
    </p>
  </div>
</div>
