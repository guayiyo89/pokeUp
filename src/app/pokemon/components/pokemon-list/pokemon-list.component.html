<div class="container">
  <div class="content">
    <h1>Buscar pokemon</h1>
    <div class="botones">
      <input
        placeholder="buscar por nombre"
        type="text"
        id="searchPokemon"
        [formControl]="searchPokemon"
        (keyup.enter)="showPokemon()"
      />
    </div>

    <div class="botones ptop12">
      <button id="buscar" (click)="showPokemon()">Buscar</button>
      <button id="restart" (click)="loadPokemons()">Reestablecer</button>
    </div>

    <div class="content-list">
      <div
        class="pokemon-listed"
        (click)="openModal(pokemon)"
        *ngFor="let pokemon of currentPageResults"
      >
        {{ pokemon.name | titlecase }}
      </div>
      <div class="msg-alerta" *ngIf="currentPageResults.length == 0">
        No hay resultados que coincidan.
      </div>
    </div>

    <div class="pagination" *ngIf="currentPageResults.length != 0">
      <button id="btnInicio" (click)="showFirstPage()">Inicio</button>
      <button id="btnAtras" (click)="beforePage()"><<</button>
      {{ currentPage }} / {{ lastPage }}
      <button id="btnSgte" (click)="nextPage()">>></button>
      <button id="btnFin" (click)="showLastPage()">Final</button>
    </div>
  </div>
</div>

<div class="modal_class_open" *ngIf="modalSwitch == true">
  <app-modal-pokemon></app-modal-pokemon>
</div>
