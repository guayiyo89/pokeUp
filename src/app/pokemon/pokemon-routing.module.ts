import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { NotFoundComponent } from '../shared/components/not-found/not-found.component';
import { PokemonListComponent } from './components/pokemon-list/pokemon-list.component';
import { PokepediaComponent } from './components/pokepedia/pokepedia.component';

const routes: Routes = [
  {
    path: '',
    pathMatch: 'full',
    redirectTo: '/buscar-pokemon',
  },
  {
    path: 'buscar-pokemon',
    component: PokemonListComponent,
    children: [],
  },
  {
    path: 'pokepedia',
    component: PokepediaComponent,
    children: [],
  },
  { path: '404', component: NotFoundComponent },
  { path: '**', redirectTo: '/404' },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class PokemonRoutingModule {}
