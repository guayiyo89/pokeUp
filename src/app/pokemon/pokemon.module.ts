import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PokemonListComponent } from './components/pokemon-list/pokemon-list.component';
import { PokemonRoutingModule } from './pokemon-routing.module';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { ModalPokemonComponent } from './components/modal-pokemon/modal-pokemon.component';

@NgModule({
  declarations: [
    PokemonListComponent,
    ModalPokemonComponent
  ],
  imports: [
    CommonModule,
    FormsModule,
    ReactiveFormsModule,
    PokemonRoutingModule
  ]
})
export class PokemonModule { }
