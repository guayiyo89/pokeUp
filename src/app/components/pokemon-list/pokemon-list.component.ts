import { Component, OnInit } from '@angular/core';
import { FormControl } from '@angular/forms';
import { Base } from 'src/app/interfaces/base';
import { PokemonService } from 'src/app/services/pokemon.service';

@Component({
  selector: 'app-pokemon-list',
  templateUrl: './pokemon-list.component.html',
  styleUrls: ['./pokemon-list.component.css'],
})
export class PokemonListComponent implements OnInit {
  listPokemon: Base[] = [];
  load = false;
  searchPokemon = new FormControl('');
  pageActual: Base[] = [];
  pageSize = 5;
  lastPage = 1;
  currentPage = 1;

  constructor(private pokeSvc: PokemonService) {}

  ngOnInit(): void {
    this.getPokemons();
  }

  getPokemons() {
    this.pokeSvc.getPokemons1stGen().subscribe((resp) => {
      this.listPokemon = resp.pokemon_species;
      this.lastPage = this.calcularPaginas(this.listPokemon, this.pageSize);
      this.pageActual = this.listPokemon;
      this.load = true;
    });
  }

  showPokemon() {
    this.pageActual = this.listPokemon.filter((specie) =>
      specie.name.includes(this.searchPokemon.value)
    );
  }

  paginate(array: Base[], page_size: number, page_number: number) {
    return array.slice((page_number - 1) * page_size, page_number * page_size);
  }

  calcularPaginas(array: Base[], page_size: number){
    let totalPaginas;
    if(array.length % page_size == 0){
      totalPaginas = array.length / page_size
    } else {
      totalPaginas = Math.trunc(array.length / page_size) + 1;
    }
    return totalPaginas
  }
  
}
