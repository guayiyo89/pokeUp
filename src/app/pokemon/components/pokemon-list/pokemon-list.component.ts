import { Component, OnInit } from '@angular/core';
import { FormControl } from '@angular/forms';
import { ModalService } from 'src/app/shared/modal.service';
import { PassingDataService } from 'src/app/shared/passing-data.service';
import { Base } from '../../interfaces/base';
import { PokemonService } from '../../services/pokemon.service';

@Component({
  selector: 'app-pokemon-list',
  templateUrl: './pokemon-list.component.html',
  styleUrls: ['./pokemon-list.component.css'],
})
export class PokemonListComponent implements OnInit {
  listPokemon: Base[] = [];
  results: Base[] = [];
  currentPageResults: Base[] = [];
  searchPokemon = new FormControl('');
  pageSize = 8;
  lastPage = 1;
  currentPage = 1;
  modalSwitch = false

  constructor(private pokeSvc: PokemonService, public modalSvc: ModalService, private dataSvc: PassingDataService) {
  }

  ngOnInit(): void {
    this.loadPokemons();
    this.modalSvc.$modal.subscribe((value) => this.modalSwitch = value);
  }

  loadPokemons() {
      this.listPokemon = this.pokeSvc.dataPokemon;
      this.lastPage = this.calcularPaginas(this.pokeSvc.dataPokemon, this.pageSize);
      this.results = this.listPokemon.sort((a, b) =>
        a.name.localeCompare(b.name)
      );
      this.currentPage = 1;
      this.currentPageResults = this.paginate(
        this.results,
        this.pageSize,
        1
      );
  }

  showPokemon() {
    this.results = this.listPokemon.filter((specie) =>
      specie.name.includes(this.searchPokemon.value)
    );
    this.lastPage = this.calcularPaginas(this.results, this.pageSize);
    this.currentPage = 1;
    this.currentPageResults = this.paginate(
      this.results,
      this.pageSize,
      this.currentPage
    );
  }

  paginate(array: Base[], page_size: number, page_number: number) {
    return array.slice((page_number - 1) * page_size, page_number * page_size);
  }

  calcularPaginas(array: Base[], page_size: number) {
    let totalPaginas;
    if (array.length % page_size == 0) {
      totalPaginas = array.length / page_size;
    } else {
      totalPaginas = Math.trunc(array.length / page_size) + 1;
    }
    return totalPaginas;
  }

  nextPage() {
    if (this.currentPage != this.lastPage) {
      this.currentPage++;
      this.currentPageResults = this.paginate(
        this.results,
        this.pageSize,
        this.currentPage
      );
    }
  }

  beforePage() {
    if (this.currentPage != 1) {
      this.currentPage--;
      this.currentPageResults = this.paginate(
        this.results,
        this.pageSize,
        this.currentPage
      );
    }
  }

  showFirstPage() {
    this.currentPageResults = this.paginate(this.results, this.pageSize, 1);
    this.currentPage = 1;
  }

  showLastPage() {
    this.currentPageResults = this.paginate(
      this.results,
      this.pageSize,
      this.lastPage
    );
    this.currentPage = this.lastPage;
  }

  openModal(pokemon: Base) {
    this.dataSvc.enviarData(pokemon)
    this.modalSwitch = true
  }


}
