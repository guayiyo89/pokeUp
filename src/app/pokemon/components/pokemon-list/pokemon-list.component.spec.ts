import { HttpClientTestingModule } from '@angular/common/http/testing';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { RouterTestingModule } from '@angular/router/testing';
import { ModalService } from '../../../../app/shared/modal.service';
import { PokemonService } from '../../services/pokemon.service';
import { FilteredCreaturesMock, PokemonCreaturesMock, SortedCreaturesMock } from '../../services/species.mock';

import { PokemonListComponent } from './pokemon-list.component';

describe('PokemonListComponent', () => {
  let component: PokemonListComponent;
  let fixture: ComponentFixture<PokemonListComponent>;
  let pokeSvc: PokemonService;
  let modalSvc: ModalService;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [RouterTestingModule, HttpClientTestingModule, ReactiveFormsModule],
      providers: [PokemonService],
      declarations: [ PokemonListComponent ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(PokemonListComponent);
    component = fixture.componentInstance;
    compiled = fixture.nativeElement;
    pokeSvc = TestBed.inject(PokemonService);
    pokeSvc.dataPokemon = PokemonCreaturesMock
    expect(pokeSvc.dataPokemon).toBeInstanceOf(Array)
    fixture.detectChanges();
  });

  afterEach(() => {
    pokeSvc.dataPokemon = PokemonCreaturesMock
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('When initializing', () => {
    test('# should call loadPokemons method', () => {
      const spy = jest.spyOn(component, 'loadPokemons');
      component.ngOnInit();
      expect(spy).toHaveBeenCalled();
    })

    test('# should get the value of modalSwitch from the service and has to be false', () => {
      modalSvc = TestBed.inject(ModalService);
      component.ngOnInit();
      modalSvc.$modal.subscribe((val) => expect(val).toBe(false))
    })

    test('# should call sortResult', () => {
      const spy = jest.spyOn(component, 'sortResult');
      component.ngOnInit();
      component.loadPokemons();
      expect(spy).toHaveBeenCalled();
    })

    test('# should call calcularPaginas', () => {
      const spy = jest.spyOn(component, 'calcularPaginas');
      component.ngOnInit();
      component.loadPokemons();
      expect(spy).toHaveBeenCalled();
    })

    test('# should call paginate', () => {
      const spy = jest.spyOn(component, 'paginate');
      component.ngOnInit();
      component.loadPokemons();
      expect(spy).toHaveBeenCalled();
    })

    describe('When loadPokemon is called', () => {
      beforeEach(() => {
        component.ngOnInit();
        component.loadPokemons();
        fixture.detectChanges();
      });

      test('# should sort the result when sortResult is called', () => {
        const sortedRes = SortedCreaturesMock
        const expectedRes = component.listPokemon
  
        component.sortResult(pokeSvc.dataPokemon);
  
        expect(expectedRes).toStrictEqual(sortedRes)
      })
  
      test('# should set the number of lastPage when calcularPaginas is called', () => {
        const expectedRes = component.lastPage
        const numPages = component.pageSize
  
        component.calcularPaginas(pokeSvc.dataPokemon, numPages)
  
        expect(expectedRes).toBe(1)
      })

      test('# should set currentPageResults for the first called', () => {
        const numPages = component.pageSize
        const sortedRes = SortedCreaturesMock
        const expectedRes = component.currentPageResults

        component.paginate(sortedRes, numPages, 1)

        expect(expectedRes).toStrictEqual(sortedRes)
      })

    })

  })

  describe('When some pokemon name of the list is clicked', () => {
    test('Should open the modal and set the status of modalSwitch to true', () => {
      const spy = jest.spyOn(component, 'openModal');
      const pokeListed = fixture.debugElement.query(By.css('.pokemon-listed'))

      pokeListed.nativeElement.click();
      fixture.detectChanges();

      expect(spy).toHaveBeenCalled;
      expect(component.modalSwitch).toBe(true)
    })
  })

  describe('When the data is loaded and sorted', () => {
    beforeEach(() => {
      component.ngOnInit();
      component.loadPokemons();
      component.listPokemon = SortedCreaturesMock;
      fixture.detectChanges();
    });

    test('# showPokemon is called when Buscar is clicked', () => {
      const spy = jest.spyOn(component, 'showPokemon');
      const button = fixture.debugElement.query(By.css('#buscar'))

      button.nativeElement.click();
      fixture.detectChanges();

      expect(spy).toHaveBeenCalled;
    })

    test('# loadPokemons is called when Buscar is clicked', () => {
      const spy = jest.spyOn(component, 'loadPokemons');
      const button = fixture.debugElement.query(By.css('#restart'))

      button.nativeElement.click();
      fixture.detectChanges();

      expect(spy).toHaveBeenCalled;
    })

    test('# set result with filtered values after pressing Buscar', () => {
      const filteredRes = FilteredCreaturesMock
      const button = fixture.debugElement.query(By.css('#buscar'))
      let result = component.results

      button.nativeElement.click();
      component.showPokemon();
      result = component.listPokemon.filter((val) => val.name.includes('iza'))
      fixture.detectChanges();

      expect(result).toStrictEqual(filteredRes);
    })

    
  })
});
