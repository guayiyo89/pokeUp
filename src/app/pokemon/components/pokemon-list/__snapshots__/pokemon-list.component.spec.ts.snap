// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonListComponent When initializing #should do match with snapshot 1`] = `
"<div class=\\"container\\"><div class=\\"content\\"><h1>Buscar pokemon</h1><div class=\\"botones\\"><input id=\\"searchPokemon\\" placeholder=\\"buscar por nombre\\" type=\\"text\\" ng-reflect-form=\\"[object Object]\\" class=\\"ng-untouched ng-pristine ng-valid\\"></div><div class=\\"botones ptop12\\"><button id=\\"buscar\\">Buscar</button><button id=\\"restart\\">Reestablecer</button></div><div class=\\"content-list\\"><!--bindings={
  \\"ng-reflect-ng-for-of\\": \\"[object Object],[object Object\\"
}--><div class=\\"pokemon-listed\\"> Abra </div><div class=\\"pokemon-listed\\"> Charizard </div><div class=\\"pokemon-listed\\"> Dragizard </div><div class=\\"pokemon-listed\\"> Eevee </div><div class=\\"pokemon-listed\\"> Meowth </div><div class=\\"pokemon-listed\\"> Pikachu </div><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}--></div><!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><div class=\\"pagination\\"><button id=\\"btnInicio\\">Inicio</button><button id=\\"btnAtras\\">&lt;&lt;</button> 1 / 1 <button id=\\"btnSgte\\">&gt;&gt;</button><button id=\\"btnFin\\">Final</button></div></div></div><!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}-->"
`;
